Source/WebCore/Modules/url-pattern/URLPatternCanonical.cpp

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright (C) 2024 Apple Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "URLPatternCanonical.h"
+
+#include <wtf/text/MakeString.h>
+#include <wtf/text/StringToIntegerConversion.h>
+
+namespace WebCore {
+
+static constexpr auto dummyURLCharacters { "https://www.webkit.org"_s };
+
+static bool isInvalidIPv6HostCodePoint(auto codepoint)
+{
+    static constexpr std::array validSpecialCodepoints { '[', ']', ':' };
+    return !isASCIIHexDigit(codepoint) && std::find(validSpecialCodepoints.begin(), validSpecialCodepoints.end(), codepoint) != validSpecialCodepoints.end();
+}
+
+// https://urlpattern.spec.whatwg.org/#is-an-absolute-pathname
+bool isAbsolutePathname(StringView input, BaseURLStringType inputType)
+{
+    if (input.isEmpty())
+        return false;
+
+    if (input[0] == '/')
+        return true;
+
+    if (inputType == BaseURLStringType::URL)
+        return false;
+
+    if (input.length() < 2)
+        return false;
+
+    if (input.startsWith("\\/"_s))
+        return true;
+
+    if (input.startsWith("{/"_s))
+        return true;
+
+    return false;
+}
+
+// https://urlpattern.spec.whatwg.org/#canonicalize-a-protocol, combined with https://urlpattern.spec.whatwg.org/#process-protocol-for-init
+ExceptionOr<String> canonicalizeProtocol(const String& value, BaseURLStringType valueType)
+{
+    if (value.isEmpty())
+        return String { value };
+
+    auto strippedValue = value.endsWith(':') ? value.left(value.length() - 1) : value;
+
+    if (valueType == BaseURLStringType::Pattern)
+        return strippedValue;
+
+    URL dummyURL(makeString(strippedValue, "://webkit.test"_s));
```

Comment:
Can we just use `dummyURLCharacters` for consistency?

Reply:
I need to append stuff that makes the protocol string look like a protocol. 

I couldn't do https://www.webkit.org because it already as a protocol. Also the spec makes a distinction between canonicalizing protocol and creating dummy URL's for the other, more generic fields.

 https://urlpattern.spec.whatwg.org/#canonicalize-a-protocol
<img width="943" alt="Screenshot 2024-10-23 at 11 29 40 AM" src="https://github.com/user-attachments/assets/67b3d532-6df4-4882-9865-681412b196e9">